libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts

```diff
@@ -1038,165 +1027,314 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
-                        for (const property in val) {
-                            const newInfo: Partial<EntityInfo> = Object.assign({}, info);
-                            newInfo.property = property;
-                            infos.push(newInfo);
+                if (entityName) {
+                    this.expandEntity(
+                        entityName,
+                        entity as Record<string, unknown[]>,
+                        instance,
+                        root,
+                        op,
+                        property,
+                        turn,
+                        text,
+                        entityToInfo
+                    );
+                } else if (typeof entity === 'object' && entity !== null) {
+                    if (isEmpty(entity)) {
+                        if (isOp) {
+                            // Handle operator with no children.
+                            this.expandEntity(op, null, instance, root, op, property, turn, text, entityToInfo);
+                        } else if (isProperty) {
+                            // Handle property with no children.
+                            this.expandEntity(property, null, instance, root, op, property, turn, text, entityToInfo);
                         }
                     } else {
-                        infos.push(info);
+                        this.expandEntityObject(
+                            entity as Record<string, unknown[]>,
+                            op,
+                            property,
+                            root,
+                            operations,
+                            properties,
+                            turn,
+                            text,
+                            entityToInfo
+                        );
                     }
+                } else if (isOp) {
+                    // Handle global operator with no children in model.
+                    this.expandEntity(op, null, instance, root, op, property, turn, text, entityToInfo);
                 }
-            }
+            });
+        }
+    }
+
+    private stripProperty(name: string): string {
+        return name.endsWith(this.propertyEnding) ? name.substring(0, name.length - this.propertyEnding.length) : name;
+    }
+
+    private expandEntity(
+        name: string,
+        value: Record<string, unknown>,
+        instance: Record<string, unknown>,
+        rootInstance: Record<string, unknown>,
+        op: string,
+        property: string,
+        turn: number,
+        text: string,
+        entityToInfo: Record<string, EntityInfo[]>
+    ): void {
+        if (instance && rootInstance) {
+            entityToInfo[name] ??= [];
+
+            const info: EntityInfo = {
+                whenRecognized: turn,
+                name,
+                value,
+                operation: op,
+                property,
+                start: <number>rootInstance.startIndex,
```

Comment:
Are these `<T>` necessary?

Reply:
Since `instance` and `rootInstance` are one of those weird nested objects, all of their properties are `unknown`. The linter complains if I remove the casting. Granted, at this point, we kind of know the shapes of `instance` and `rootInstance`; I just couldn't find a pre-defined type. This is currently [what .NET does](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs#L1184), too, not that that means we can't make it better. Let me know how you'd like me to proceed.